06_puppy/sirigithub/mapstore.go

```diff
@@ -0,0 +1,46 @@
+package main
+
+import (
+	"fmt"
+)
+
+type MapStore map[int]Puppy
+
+func NewMapStore() MapStore {
+	return MapStore{}
+}
+
+// CreatePuppy takes a  user provided puppy, and creates a new Puppy in the store
+// returns the ID of the new Puppy.
+func (m MapStore) CreatePuppy(puppy *Puppy) int {
+	puppy.ID = len(m)
+	m[puppy.ID] = *puppy
+	return puppy.ID
+}
+
+// UpdatePuppy overrides an existing puppy in the store,
+// returns an error if id is not found or does not match the puppy ID
+func (m MapStore) UpdatePuppy(p *Puppy) error {
+	if _, ok := m[p.ID]; !ok {
+		return fmt.Errorf("puppy ID %d to update does not exist in the map", p.ID)
+	}
+	m[p.ID] = *p
+	return nil
+}
+
+// DeletePuppy deletes an existing puppy from the store
+func (m MapStore) DeletePuppy(id int) error {
+	if _, ok := m[id]; ok {
```

Comment:
I would suggest inverting the sense of this test - that is, use `!ok` and return the error from inside the `if` block. Then outside the `if` you can delete the item from the map.

This is described more in: https://medium.com/@matryer/line-of-sight-in-code-186dd7cdea88 called aligning the happy path to the left edge.

Reply:
That was a very informative read thanks @camh-anz . Updated code.